scirpy/io/_io.py

```diff
@@ -392,54 +434,135 @@ def read_bracer(path: Union[str, Path]) -> AnnData:
+                row["V_SEQ_START"] + row["V_SEQ_LENGTH"]
+            )  # type: ignore
+            chain_dict["np2_length"] = row["J_SEQ_START"] - (
+                row["D_SEQ_START"] + row["D_SEQ_LENGTH"]
+            )  # type: ignore
+
+        chain_dict["junction"] = (
+            row["JUNCTION"] if not pd.isnull(row["JUNCTION"]) else None
         )
-        tmp_ir_cell.add_chain(tmp_chain)
+        chain_dict["junction_aa"] = (
+            _translate_dna_to_protein(chain_dict["junction"])
+            if chain_dict["junction"] is not None
+            else None
+        )
+        chain_dict["consensus_count"] = row["TPM"]
+        chain_dict["productive"] = row["FUNCTIONAL"]
+
+        tmp_ir_cell.add_chain(chain_dict)
 
     return from_ir_objs(bcr_cells.values())
+
+
+def write_airr(adata):
+    # TODO
+    pass
+
+
+def update_schema(adata):
+    # TODO
+    pass
+
+
+def to_dandelion(adata):
+    """
+    Convert a scirpy-initialized AnnData object to Dandelion format using `to_ir_objs`.
+
+    Parameters
+    ----------
+    adata
+        annotated data matrix with :term:`IR` annotations.
+
+    Returns
+    -------
+    `Dandelion` object.
+
+    """
+    try:
+        import dandelion as ddl
+    except:
+        raise ImportError("Please install dandelion: pip install sc-dandelion.")
+    airr_cells = to_ir_objs(adata)
+
+    contig_dicts = {}
+    for tmp_cell in airr_cells:
+        for i, tmp_chain in enumerate(tmp_cell.chains, start=1):
+            tmp_chain.update(
+                {
+                    "cell_id": tmp_cell.cell_id,
+                    "sequence_id": f"{tmp_cell.cell_id}_contig_{i}",
+                }
+            )
+            tmp_chain["umi_count"] = tmp_chain.pop("duplicate_count")
+            contig_dicts[tmp_chain["sequence_id"]] = tmp_chain
+
+    data = pd.DataFrame.from_dict(contig_dicts, orient="index")
+    return ddl.Dandelion(ddl.load_data(data))
+
+
+def from_dandelion(dandelion, transfer=False) -> AnnData:
+    """Import data from dandelion :cite:`Stephenson2021`.
+
+    Parameters
+    ----------
+    dandelion
+        a :class:`dandelion.Dandelion` instance
+    transfer
+        Whether to execute :func:`dandelion.tl.transfer` to transfer all data
+        to the :class:`anndata.AnnData` instance.
+
+    Returns
+    -------
+    A :class:`~anndata.AnnData` instance with AIRR information stored in `obs`.
+    """
+    try:
+        import dandelion as ddl
+    except ImportError:
+        raise ImportError("Please install dandelion: pip install sc-dandelion.")
+
+    dandelion_df = dandelion.data.copy()
+    # replace "unassigned" with None
+    for col in dandelion_df.columns:
+        dandelion_df.loc[dandelion_df[col] == "unassigned", col] = None
+
+    adata = read_airr(dandelion_df)
+    if transfer:
+        ddl.tl.transfer(
+            adata, dandelion
+        )  # need to make a version that is not so verbose?
+    return adata
```

Comment:
@zktuong, most of this PR is still work-in-progress, but the round trip-conversion of a 10x dataset from scirpy to dandelion and back seems to work now. Could you please have another look at these functions if everything makes sense? 

Unlike the previous `IrCell/IrChain` implementation, `AirrCell` now simply stores chains as d AIRR rearrangement-compliant dictionaries.

Reply:
A couple of hiccups. I've attached the notebook as a pdf.

[testing_scirpy_dandelion_conversion.pdf](https://github.com/icbi-lab/scirpy/files/6228705/testing_scirpy_dandelion_conversion.pdf)

One of it is on my end -> I need to switch the requirement for umi_count to allow duplicate_count as mentioned. Seems like my init method is not catching that properly.

Reply:
also it's overly verbose i think

Reply:
ah actually the issue is with 'locus'. it's doesn't come back as a column with `read_airr` (or `airr.load_rearrangement`), although it's a standard term (but optional according to airr). It's a hard requirement for dandelion.

Reply:
Thanks for having a look! In your PR you wrote a function to infer the locus name from the genes (see https://github.com/icbi-lab/scirpy/pull/241/files#diff-5e72138c27e2f83f0219586fad1d67885a1b16ceaf3e327447bba0f5ab1f85d9R379). That's certainly a useful feature and I'm adding it to `read_airr` now.

Reply:
Thanks @grst. That fixed it.

I encountered a weird bug that i will try to fix on dandelion's side first. This has got to do with empty `clone_id` columns in the airr table. scirpy passes that empty entry into `AirrCell` fine but then this messes with dandelion's initialization procedure but a simple fix will correct that. 

It did also revealed that if i were to call clones from scirpy and try to convert to dandelion, this information is not passed on and meets the same issue above. Is it because in scirpy `clone_id` is called `clonotype`?

Reply:
Currently only the chain-specific attributes get exported (i.e. all scirpy columns that start with `IR_`). 
In principle, it probably makes sense to write out the `clonotype` column to `clone_id`. But then again, scirpy allows 
different versions of clonal assignments...

Maybe something for the next PR :shrug:

Reply:
OK. sounds good. I will just add a small conversion on dandelion's io from scirpy to parse the `clonotype` column to `clone_id` as a default, unless otherwise specified which the user can adjust.

Reply:
See #258